File: src/main/linkedlist/java/Exec.java
import java.io.*;
import java.util.Arrays;

public class Exec {
    public static void main(String[] args) {
        try {
            // Definir os métodos a serem testados
            String[] methods = { "addFirst", "add", "addLast",
                                  "getFirst", "get", "getLast",
                                  "removeFirst", "remove", "removeLast" };

            // Lendo o arquivo de entrada
            BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
            String line;

            while ((line = reader.readLine()) != null) {
                if (line.isBlank()) continue;

                int[] input = Arrays.stream(line.split(" ")).mapToInt(Integer::parseInt).toArray();

                for (String method : methods) {
                    long[] results = new long[30];

                    for (int i = 0; i < 30; i++) {    
                        LinkedList ll = new LinkedList();

                        // Adicionando os valores na LinkedList
                        for (String num: line.split(" "))
                            ll.addLast(Integer.parseInt(num));

                        long start = 0;
                        long end = 0;

                        switch (method) {
                            case "addFirst":
                                start = System.nanoTime();
                                ll.addFirst(10);
                                end = System.nanoTime();
                                break;

                            case "add":
                                start = System.nanoTime();
                                ll.add(10, input.length / 2);
                                end = System.nanoTime();
                                break;

                            case "addLast":
                                start = System.nanoTime();
                                ll.addLast(10);
                                end = System.nanoTime();
                                break;

                            case "getFirst":
                                start = System.nanoTime();
                                ll.getFirst();
                                end = System.nanoTime();
                                break;

                            case "get":
                                start = System.nanoTime();
                                ll.get(input.length / 2);
                                end = System.nanoTime();
                                break;

                            case "getLast":
                                start = System.nanoTime();
                                ll.getLast();
                                end = System.nanoTime();
                                break;

                            case "removeFirst":
                                start = System.nanoTime();
                                ll.removeFirst();
                                end = System.nanoTime();
                                break;

                            case "remove":
                                start = System.nanoTime();
                                ll.removeByIndex(input.length / 2);
                                end = System.nanoTime();
                                break;

                            case "removeLast":
                                start = System.nanoTime();
                                ll.removeLast();
                                end = System.nanoTime();
                                break;
                        }

                        results[i] = (end - start);
                    }

                    Arrays.sort(results); // Ordena para buscar a mediana
                    String outputLine = "linkedlist-java " + results[14] + " " + input.length;

                    // Define o arquivo de saída baseado no método
                    File file = new File(method + ".data");
                    boolean isNewFile = !file.exists() || file.length() == 0;

                    try (PrintWriter writer = new PrintWriter(new FileOutputStream(file, true))) {
                        if (isNewFile) {
                            writer.println("estrutura-linguagem tempo tamanho_da_entrada");
                        }
                        writer.println(outputLine);
                    }
                }
            }
        } catch (IOException ioe) {
            ioe.printStackTrace();
        }
    }
}

File: src/main/linkedlist/python/Exec.py
import sys
import time
import os
import statistics
from linked_list import LinkedList

methods = [
    "addFirst", "add", "addLast",
    "getFirst", "get", "getLast",
    "removeFirst", "remove", "removeLast"
]

def mede_tempo(ll, method, n):
    index = n // 2
    start = end = 0

    if method == "addFirst":
        start = time.perf_counter_ns()
        ll.add_first(10)
        end = time.perf_counter_ns()
        ll.remove_first()

    elif method == "add":
        start = time.perf_counter_ns()
        ll.add(10, index)
        end = time.perf_counter_ns()
        ll.remove_by_index(index)

    elif method == "addLast":
        start = time.perf_counter_ns()
        ll.add_last(10)
        end = time.perf_counter_ns()
        ll.remove_last()

    elif method == "getFirst":
        start = time.perf_counter_ns()
        ll.get_first()
        end = time.perf_counter_ns()

    elif method == "get":
        start = time.perf_counter_ns()
        ll.get(index)
        end = time.perf_counter_ns()

    elif method == "getLast":
        start = time.perf_counter_ns()
        ll.get_last()
        end = time.perf_counter_ns()

    elif method == "removeFirst":
        val = ll.get_first()
        start = time.perf_counter_ns()
        ll.remove_first()
        end = time.perf_counter_ns()
        ll.add_first(val)

    elif method == "remove":
        val = ll.get(index)
        start = time.perf_counter_ns()
        ll.remove_by_index(index)
        end = time.perf_counter_ns()
        ll.add(val, index)

    elif method == "removeLast":
        val = ll.get_last()
        start = time.perf_counter_ns()
        ll.remove_last()
        end = time.perf_counter_ns()
        ll.add_last(val)

    return end - start


def salva_resultado(method, tempo, tamanho):
    file_name = f"{method}.data"
    is_new_file = not os.path.exists(file_name) or os.stat(file_name).st_size == 0

    with open(file_name, "a") as file:
        if is_new_file:
            file.write("estrutura-linguagem tempo tamanho_da_entrada\n")
        file.write(f"linkedlist-python {tempo} {tamanho}\n")


def main():
    for line in sys.stdin:
        line = line.strip()
        if not line:
            continue

        input_list = list(map(int, line.split()))
        n = len(input_list)

        ll = LinkedList()
        for num in input_list:
            ll.add_last(num)

        for method in methods:
            resultados = []
            for _ in range(101):
                tempo = mede_tempo(ll, method, n)
                resultados.append(tempo)

            mediana = statistics.median(resultados)
            salva_resultado(method, int(mediana), n)


if __name__ == "__main__":
    main()
